app/gui/src/dashboard/hooks/searchParamsStateHooks.ts

```diff
@@ -49,16 +49,28 @@ export function useSearchParamsState<T = unknown>(
         | ((currentSearchParams: URLSearchParams) => URLSearchParams),
       options: RouteLocationOptions,
     ) => {
-      const params = searchParams
+      // We get window.location.search, because we ensure it's up-to-date. See comment below.
+      const params = new URLSearchParams(window.location.search)
 
       if (nextSearchParams instanceof Function) {
         nextSearchParams = nextSearchParams(params)
       }
 
       const query = Object.fromEntries(nextSearchParams.entries())
-      void router.push({ query, ...options })
+      // TODO[ao]: router.push/router.replace are asynchronous, but we want window.location.href
+      // to be updated immediately, so any subsequent query changes won't override this one.
+      //
+      // This keeps the old way of doing (before #12803), but I think the more elegant solution
+      // would be keeping the "intermediate" state by ourselves and leave window.location.search
+      // management to router.
+      if (options.replace ?? false) {
+        window.history.replaceState(null, '', `?${nextSearchParams.toString()}`)
```

Comment:
This can't be a solution.

Reply:
What exactly cannot be a solution?

If "using window.location.search" to keep state of current query, then it is how it worked before: https://github.com/enso-org/enso/blob/7b7890dc5e0c76e10516a324b3f3a2679731600c/app/gui/src/dashboard/hooks/searchParamsStateHooks.ts#L56 It just relied on navigate updating window.location.search synchronously, while now it is asynchronous.

I may implement better solution, not relying on window.location.search (as in TODO), but wanted to merge a quick fix first.

Reply:
Could you add a reference to a code where we update window.location directly?

Reply:
> Could you add a reference to a code where we update window.location directly?

There is no such code. Please, re-read my message carefully: I said _keeping_ state of current query. Previous code was relying on `navigate` setting `window.location.search` synchronously. This is not something guaranteed, according to my research (but I might be wrong).

Reply:
In react-router it's always synchronous, same true for `window.replaceState(...)`. In the code you mentioned it serves a different purpose - in some edge cases we can have an outdated value in react (because react update states in batch), see: https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state

react-router doesn't provide an option to get the current pending value so we need to read the location directly.

---

My initial comment was about the workaround with `window.history.replaceState` and it's _obviously_ not a solution and I won't let it merged that way.

Reply:
1. It's indirect.
2. It happens outside of the router context
3. It runs states out-of-sync
4. It's not ideomatic.
5. This is clearly a hack
6. I spent the last year getting rid of code like this.